### PR TITLE
feat: support crate imports in deno_typescript

### PR DESCRIPTION
**Pull request by afinch7**
Basic support for crate imports for `deno_typescript`. Syntax is like this `crate://crate_name/path_in_crate.ts`.

### CODE REVIEW

File: cli/js/deno.ts [hunk 1]
```diff
@@ -118,3 +118,4 @@ export let pid: number;
 export let noColor: boolean;
 
 export { symbols } from "./symbols.ts";
+export { testFunction } from "crate://deno_core/test_file.ts";
```
Comment (ry) [inline cli/js/deno.ts]: This is pretty cool.

File: core/lib.rs [hunk 2]
```diff
@@ -45,3 +46,5 @@ pub fn v8_version() -> &'static str {
 fn test_v8_version() {
   assert!(v8_version().len() > 3);
 }
+
+crate_modules!["test_file.ts"];
```
Comment (ry) [inline core/lib.rs]: Would be nice if it wasn't required to list them like this. Not sure if that's possible to avoid tho.

Comment (bartlomieju) [reply core/lib.rs]: Could `crate_modules!`/`include_crate_modules!` return path to the installed crate? That way, we could easily resolve arbitrary files from crate, without needing to specify them explicitly.

Comment (afinch7) [reply core/lib.rs]: @bartlomieju 👍 I like that better. Still don't think it's perfect, but it makes things much simpler.

File: cli/js/deno.ts (hunk 1 above)
Comment (bartlomieju) [reply cli/js/deno.ts]: @afinch7 can you convert it to `import` so it's not exposed?

File: core/modules.rs [hunk 3]
```diff
@@ -449,6 +449,26 @@ impl fmt::Display for Deps {
...
+  ( $( $x:ident ),* ) => {
+    {
+      let mut temp: HashMap<String, String> = HashMap::new();
+      $(
+        temp.insert(stringify!($x).to_string(), $x::DENO_CRATE_PATH.to_string());
+      )*
+      temp
+    }
+  };
+}
```
Comment (bartlomieju) [inline core/modules.rs]: This solution is so simple that these macros almost feel like extraneous sugar 👍

File: cli/js/deno.ts (hunk 1 above)
Comment (afinch7) [reply cli/js/deno.ts]: Seems to work fine as well, but I will remove it now that I know that it works.
```
./target/debug/deno                                                                                                            ░▒▓ ✔  1m 24s 
> Deno.testFunction("test");
Hello from testFunction: test
```

**Comment by afinch7** [comment]
@bartlomieju I'm going to refactor #3692 to use this one it lands.

**Comment by ry** [comment]
Sorry @afinch7 I guess the PR doesn't use `include_str!` ... did a previous version or am I just mistaken?

Anyway the solution looks good now that I've seen it clearly. I wonder if this works with cargo package correctly?

**Comment by ry** [comment]
Practical example:
```
import { makeTempDir } from "crate://deno_fs/make_temp_dir.ts";
```

**Comment by ry** [comment]
We should really add an explanation to the manual.

**Comment by ry** [comment]
Just need some a minimal manual entry and then we can land it

**Comment by bartlomieju** [comment]
@ry this is snapshot only feature, does it really need entry in manual right now? Maybe let's refactor some crates and stabilize this feature first?

**Comment by ry** [comment]
Alright - you're right - let's land it.

**Pull request merged**